oap-ape/ape-java/ape-common/src/main/java/com/intel/ape/service/netty/NettyMessage.java

```diff
@@ -610,42 +615,79 @@ void write(ChannelOutboundInvoker out, ChannelPromise promise, ByteBufAllocator
 
             // compute lengths of header and content: sequenceId(4), hasNextBatch(1),
             // columnCount(4), rowCount(4), dataBufferLengths(columnCount * 4),
-            // compositeFlags(columnCount), compositedElementCount(4)
-            int headerLength = 4 + 1 + 4 + 4 + columnCount * 4 + columnCount + 4;
-            int contentLength = 0;
-            contentLength += compositedElementLengths.readableBytes();
+            // compositeFlags(columnCount), compositedElementCount(4), compressEnabled(1),
+            // contentLength(4)
+            int headerLength = 4 + 1 + 4 + 4 + columnCount * 4 + columnCount + 4 + 1 + 4;
+            int originLength = 0;
+            originLength += compositedElementLengths.readableBytes();
 
             for (int buffLength : dataBufferLengths) {
-                contentLength += buffLength;
+                originLength += buffLength;
             }
-            contentLength += (columnCount * rowCount); // for null buffers
-
-            // set header
-            ByteBuf headerBuf = allocateBuffer(allocator, ID, headerLength, contentLength, false);
-            headerBuf.writeInt(sequenceId);
-            headerBuf.writeBoolean(hasNextBatch);
-            headerBuf.writeInt(columnCount);
-            headerBuf.writeInt(rowCount);
-            for (int dataBufferLength : dataBufferLengths) {
-                headerBuf.writeInt(dataBufferLength);
-            }
-            for (boolean isComposited : compositeFlags) {
-                headerBuf.writeBoolean(isComposited);
-            }
-            headerBuf.writeInt(compositedElementCount);
-
-            // write to channel
-            out.write(headerBuf);
-            out.write(compositedElementLengths);
-            for (int i = 0; i < columnCount; i++) {
-                out.write(dataBuffers[i]);
-            }
-            if (columnCount > 0) {
+            originLength += (columnCount * rowCount); // for null buffers
+            if (!compressEnabled) {
+                // set header

```

Comment:
possible to refactor this part as a method?

Reply:
Maybe I didn't make it clear. What I expected is something like
`if(compressEnabled) {
    doCompressThings();
} else {
   doNonCompressThings(); 
}`

Reply:
updated.